### ISSUE TITLE
LiveRide gets stuck in "Replanning the journey" state

### ISSUE THREAD

**Issue by oliverlockwood** [cyclestreets/android#386]
Although the image below doesn't fully show the issue, it is an example where it occurred.
While riding I declined to take the planned route turning onto Hervines Road (which would require going on paths, and I wanted to stick to roads); the system clearly replanned just fine and updated the map as shown, but the notification stayed stuck on "Replanning the journey".  It stayed in this state even as I approached and passed the waypoint, with the location icon showing my movement throughout.

<img src="https://user-images.githubusercontent.com/6017680/81088754-57d51180-8ef3-11ea-9d14-53473d451923.jpeg" width=400>

It seems like we've introduced a bug in the state machine.

**Comment by jezhiggins** [comment]
Can't immediately reproduce - had a small play with it on my way back home from my dog walk, and replanned ok. Did seem so to pick up the segment I was on, so perhaps related?

**Comment by oliverlockwood** [comment]
I went out for my lockdown-sanctioned exercise slot this morning, running the latest beta build, and a replan worked fine. So maybe our other changes have mysteriously fixed this. I'll keep an eye out in the next few days and see if it reproduces again.

Can you still reproduce this @jezhiggins, or has it gone away for you too?

**Comment by jezhiggins** [comment]
Forced replans two or three times this morning, and worked with out issue. What might help is announcing that, not only are we about to replan the journey, but that it has actually been replanned. 

I think you could get stuck replanning if you're not somewhere reachable by the roads network - perhaps in the middle of park (or simply somewhere that isn't very well mapped) - or if you don't have a good gps lock and your position isn't certain. Neither of these cases seems especially likely, but maybe we could look at some kind of backing out.

**Comment by jezhiggins** [comment]
I should perhaps raise this as a separate issue - when replanning, we should look at dropping intermediate waypoints that we have already passed. Right now, I think they all get sent across, so we replan from whereever you are to the first waypoint, and so on. If you're 40km into a 50km loop, that's probably not what you want :)

**Comment by mvl22** [comment]
I'm not sure what the behaviour is, but surely replanning should take place in the background, and only replace the route if it come back with a result, rather than cancelling the current route and _then_ replanning.

**Comment by mvl22** [comment]
> I should perhaps raise this as a separate issue - when replanning...

Yes, please.

**Comment by oliverlockwood** [comment]
@jezhiggins @mvl22 no need to raise that issue - I already fixed it in either 3.6 or 3.7 (can't remember which). If you discover it not working, please shout, but I'm 99% sure this aspect works as it should.

**Comment by jezhiggins** [comment]
I'll double check. This was a based on being 40km into a 50km ride 🙂On 10 May 2020 17:35, Oliver Lockwood <notifications@github.com> wrote:
@jezhiggins @mvl22 no need to raise that issue - I already fixed it in either 3.6 or 3.7 (can't remember which). If you discover it not working, please shout, but I'm 99% sure this aspect works as it should.

**Comment by oliverlockwood** [comment]
Better do another 50k then :wink:

**Comment by jezhiggins** [comment]
@oliverlockwood You're right - it is replanning intermediate waypoints properly and I was mistaken.

**Comment by oliverlockwood** [comment]
Cannot reproduce any more.  Closing.

**Comment by oliverlockwood** [comment]
Actually I might try again to repro this - by turning off my data connectivity to simulate losing reception while on the go - and see what happens.

**Comment by jezhiggins** [comment]
Did this experiment this morning. As you go off route, it tries to replan, fails, and then that's it. Even if you return to the route it doesn't pick it back up. 

The liveride state machine just doesn't accomodate this. If you replan successfully, it goes back to hunting for the segment, but if replanning fails it does nothing and you're stuck.
